src/Generator.h

```diff
@@ -115,16 +126,156 @@ class GeneratorParamBase {
 public:
     EXPORT explicit GeneratorParamBase(const std::string &name);
     EXPORT virtual ~GeneratorParamBase();
-    virtual void from_string(const std::string &value_string) = 0;
-    virtual std::string to_string() const = 0;
 
     const std::string name;
 
+protected:
+    friend class GeneratorBase;
+
+    virtual void set_from_string(const std::string &value_string) = 0;
+    virtual std::string to_string() const = 0;
+
 private:
     explicit GeneratorParamBase(const GeneratorParamBase &) = delete;
     void operator=(const GeneratorParamBase &) = delete;
 };
 
+// select_type<> is to std::conditional as switch is to if:
+// it allows a multiway compile-time type definition via the form
+//
+//    select_type<cond<condition1, type1>,
+//                cond<condition2, type2>,
+//                ....
+//                cond<conditionN, typeN>>::type
+//
+// Note that the conditions are evaluated in order; the first evaluating to true
+// is chosen.
+//
+// Note that if no conditions evaluate to true, the resulting type is illegal
+// and will produce a compilation error. (You can provide a default by simply
```

Comment:
Or it will fail to substitute for SFINAE. I.e. it is the right behavior for the abstraction, but there are legitimate uses for an incomplete list not filling in.